sonar-core/src/main/java/org/sonar/core/issue/tracking/Tracker.java

```diff
@@ -0,0 +1,299 @@
+    if (!tracking.isComplete()) {
+      new BlockRecognizer<RAW, BASE>().match(rawInput, baseInput, tracking);
+    }
+  }
+
+  private void match(Tracking<RAW, BASE> tracking, SearchKeyFactory factory) {
+    if (tracking.isComplete()) {
+      return;
+    }
+
+    Multimap<SearchKey, BASE> baseSearch = ArrayListMultimap.create();
+    for (BASE base : tracking.getUnmatchedBases()) {
+      baseSearch.put(factory.create(base), base);
+    }
+
+    Collection<RAW> trackedRaws = new ArrayList<>();
+    for (RAW raw : tracking.getUnmatchedRaws()) {
+      SearchKey rawKey = factory.create(raw);
+      Collection<BASE> bases = baseSearch.get(rawKey);
+      if (!bases.isEmpty()) {
+        // TODO taking the first one. Could be improved if there are more than 2 issues on the same line.
+        // Message could be checked to take the best one.
+        BASE match = bases.iterator().next();
+        tracking.associateRawToBase(raw, match);
+        baseSearch.remove(rawKey, match);
+        trackedRaws.add(raw);
+      }
+    }
+    tracking.markRawsAsAssociated(trackedRaws);
+  }
+
+  private void relocateManualIssues(Input<RAW> rawInput, Tracking<RAW, BASE> tracking) {
+    Iterable<BASE> manualIssues = from(tracking.getUnmatchedBases()).filter(IsManual.INSTANCE);
+    for (BASE base : manualIssues) {
+      if (base.getLine() == null) {
+        // no need to relocate. Location is unchanged.
+        tracking.associateManualIssueToLine(base, 0);
+      } else {
+        String lineHash = base.getLineHash();
+        if (!Strings.isNullOrEmpty(lineHash)) {
+          int[] rawLines = rawInput.getLineHashSequence().getLinesForHash(lineHash);
+          if (rawLines.length == 1) {
+            tracking.associateManualIssueToLine(base, rawLines[0]);
+          } else if (rawLines.length == 0 && base.getLine() <= rawInput.getLineHashSequence().length()) {
+            // still valid (???). We didn't manage to correctly detect code move, so the
+            // issue is kept at the same location, even if code changes
+            tracking.associateManualIssueToLine(base, base.getLine());
+          }
+        }
+      }
+    }
+  }
+
+  private enum IsManual implements Predicate<Trackable> {
+    INSTANCE;
+    @Override
+    public boolean apply(Trackable input) {
+      return input.getRuleKey().isManual();
+    }
+  }
+
+  private interface SearchKey {
+  }
+
+  private interface SearchKeyFactory {
+    SearchKey create(Trackable trackable);
+  }
+
+  private static class LineAndLineHashKey implements SearchKey {
+    private final RuleKey ruleKey;
+    private final String lineHash;
+    private final Integer line;
+
+    LineAndLineHashKey(Trackable trackable) {
+      this.ruleKey = trackable.getRuleKey();
+      this.line = trackable.getLine();
+      this.lineHash = StringUtils.defaultString(trackable.getLineHash(), "");
+    }
+
+    @Override
+    public boolean equals(@Nonnull Object o) {
```

Comment:
![BLOCKER](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-blocker.png) Add a type test to this method. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](http://nemo.sonarqube.org/coding_rules#rule_key=squid%3AS2097)